File: frontend/src/types/returPenjualan.ts
export interface DetailReturPenjualanCreate {

  barang_id: number;

  qty: number;

  harga: number;

}


export type JenisRefund =
  | "CASH"
  | "TRANSFER"
  | "PIUTANG";


export interface ReturPenjualanCreate {

  penjualan_id: number;

  pelanggan_id?: number | null;

  alasan?: string | null;

  jenis_refund?: JenisRefund;

  detail: DetailReturPenjualanCreate[];

}


export interface ReturPenjualanDetail {

  id: number;

  retur_id: number;

  barang_id: number;

  qty: number;

  harga: number;

  subtotal: number;

  kondisi?: string | null;

}


export interface ReturPenjualan {

  id: number;

  no_retur: string;

  tanggal: string;

  penjualan_id: number;

  pelanggan_id: number;

  created_by: number;

  total: number;

  status: string;

  alasan?: string | null;

  jenis_refund: JenisRefund;

  detail: ReturPenjualanDetail[];

}


export interface ReturPenjualanPaginationResponse {

  items: ReturPenjualan[];

  total: number;

  page: number;

  size: number;

  pages: number;

}

File: backend/app/schemas/retur_penjualan.py
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ==========================================================
# DETAIL RETUR PENJUALAN
# ==========================================================

class DetailReturPenjualanCreate(BaseModel):

    barang_id: int = Field(
        gt=0
    )

    qty: int = Field(
        gt=0
    )

    harga: Decimal = Field(
        ge=0
    )

    kondisi: Optional[str] = None


# ==========================================================
# CREATE RETUR PENJUALAN
# ==========================================================

class ReturPenjualanCreate(BaseModel):

    penjualan_id: int = Field(
        gt=0
    )

    pelanggan_id: Optional[int] = None

    alasan: Optional[str] = None

    jenis_refund: str = "CASH"

    detail: List[
        DetailReturPenjualanCreate
    ]


# ==========================================================
# UPDATE RETUR PENJUALAN
# ==========================================================

class ReturPenjualanUpdate(BaseModel):

    alasan: Optional[str] = None

    jenis_refund: Optional[str] = None


# ==========================================================
# DETAIL RESPONSE
# ==========================================================

class DetailReturPenjualanResponse(BaseModel):

    id: int

    retur_id: int

    barang_id: int

    qty: int

    harga: Decimal

    subtotal: Decimal

    kondisi: Optional[str] = None

    model_config = ConfigDict(
        from_attributes=True
    )


# ==========================================================
# RESPONSE
# ==========================================================

class ReturPenjualanResponse(BaseModel):

    id: int

    no_retur: Optional[str] = None

    penjualan_id: int

    pelanggan_id: Optional[int] = None

    created_by: int

    total: Decimal

    status: str

    alasan: Optional[str] = None

    jenis_refund: str

    detail: List[
        DetailReturPenjualanResponse
    ] = []

    model_config = ConfigDict(
        from_attributes=True
    )
# ==========================================================
# PAGINATION RESPONSE
# ==========================================================

class ReturPenjualanPaginationResponse(BaseModel):
    items: List[ReturPenjualanResponse]
    total: int
    page: int
    size: int